xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,256 @@
+type clusterNode struct {
+	// A way to cancel the watch for the cluster.
+	cancelFunc func()
+
+	// A list of children, as the Node can be an aggregate Cluster.
+	children []*clusterNode
+
+	// A ClusterUpdate in order to build a list of cluster updates for CDS to send down to child
+	// XdsClusterResolverLoadBalancingPolicy.
+	clusterUpdate xdsclient.ClusterUpdate
+
+	// This boolean determines whether this Node has received an update or not. This isn't the best practice,
+	// but this will protect a list of Cluster Updates from being constructed if a cluster in the tree has not received
+	// an update yet.
+	receivedUpdate bool
+
+	clusterHandler *clusterHandler
+}
+
+// CreateClusterNode creates a cluster node from a given clusterName. This will also start the watch for that cluster.
+func createClusterNode(clusterName string, xdsClient xdsClientInterface, topLevelHandler *clusterHandler) *clusterNode {
+	c := &clusterNode{
+		clusterHandler: topLevelHandler,
+	}
+	// Communicate with the xds client here.
+	c.cancelFunc = xdsClient.WatchCluster(clusterName, c.handleResp)
+	return c
+}
+
+// This function cancels the cluster watch on the cluster and all of it's children.
+func (c *clusterNode) delete() {
+	c.cancelFunc()
+	for _, child := range c.children {
+		child.delete()
+	}
+}
+
+// Construct cluster update (potentially a list of ClusterUpdates) for a node.
+func (c *clusterNode) constructClusterUpdate() ([]xdsclient.ClusterUpdate, error) {
+	// If the cluster has not yet received an update, the cluster update is not yet ready.
+	if !c.receivedUpdate {
+		return nil, errNotReceivedUpdate
+	}
+
+	// Base case - LogicalDNS or EDS. Both of these cluster types will be tied to a single ClusterUpdate.
+	if c.clusterUpdate.ClusterType != xdsclient.ClusterTypeAggregate {
+		return []xdsclient.ClusterUpdate{c.clusterUpdate}, nil
+	}
+
+	// If an aggregate construct a list by recursively calling down to all of it's children.
+	var childrenUpdates []xdsclient.ClusterUpdate
+	for _, child := range c.children {
+		childUpdateList, err := child.constructClusterUpdate()
+		if err != nil {
+			return nil, err
+		}
+		childrenUpdates = append(childrenUpdates, childUpdateList...)
+	}
+	return childrenUpdates, nil
+}
+
+// deltaInClusterUpdate determines whether there was a delta in the cluster update "fields" (meaning irrespective of children)
+// received from the xdsclient vs. the clusterUpdate that was currently present for a given cluster node. This will be used to
+// help decide whether to ping the cluster handler at the end of the handleResp() callback or not.
+func deltaInClusterUpdateFields(clusterUpdateReceived xdsclient.ClusterUpdate, clusterUpdateCurrent xdsclient.ClusterUpdate) bool {
+	if clusterUpdateReceived.ClusterName != clusterUpdateCurrent.ClusterName {
+		return true
+	}
+	if clusterUpdateReceived.EnableLRS != clusterUpdateCurrent.EnableLRS {
```

Comment:
Actually, the CDS balancer doesn't need to compare those fields `EnableLRS`, or `SecurityCfg`.

The cluster_resolver balancers will start a watch for all the leaf notes again, and deal with those fields. So the CDS balancer only needs to track `ClusterType` and `ClusterName`.

Reply:
Hmm, thinking more about it, the ClusterName can never change (the ClusterName in the response is a copy of the one in the request, and we would create new nodes if the requested cluster name changes). So the only thing that would change is the Type, and that's also the only thing that we care about.